### PR TITLE
Evals: performance improvement for quality scoring subagent + show topics grouped by area in eval scope prompt

### PR DESCRIPTION
**Pull request by nehaoss**
**Why**: Two separate improvements bundled together:

1. When asking users what to generate, the agent showed a flat topic count with no context about what integrations exist — making it hard to pick a subset. Grouping was inconsistent because it wasn't instructed.

2. Quality validation ran entirely through the Copilot agent chain (~2-5 min). The existing [evaluate_evals.py](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8b640eef5a/resources/app/out/vs/code/electron-browser/workbench/workbench.html) script does the same scoring via direct API in seconds — it just wasn't wired in.

**What changed:**

**create/SKILL.md — topic grouping in scope prompt:**

- Step 1 now explicitly groups user-facing topics by area, derived from common words in topic file names. No hardcoded package list — works for any agent.
- Both scope prompts (fresh agent + partial coverage) always render the grouped topic list before asking what to generate.
- Scope question now offers "a specific area" as an explicit option alongside "full evaluation" and "specific category."
- Uses BeginDialog target as a tiebreaker when file names are ambiguous.

**validate/SKILL.md — script-first hybrid validation:**

- Step 1: run [evaluate_evals.py --category {category}](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8b640eef5a/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (fast, ~5 sec via Copilot API).
- Step 2: manual scoring fallback — only if script fails (auth error, missing dependency).
- Report announces which path was used: 📊 Scored using evaluate_evals.py or ⚠️ Script unavailable — scored manually.

**[quality-fix-flow.md](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8b640eef5a/resources/app/out/vs/code/electron-browser/workbench/workbench.html) — targeted re-validation:**

- Re-validation after fixes now passes only the edited files and only the flagged dimensions — skips unchanged files and unaffected dimensions.

- [evaluate_evals.py](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8b640eef5a/resources/app/out/vs/code/electron-browser/workbench/workbench.html) — script improvements:

**[evaluate_evals.py](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8b640eef5a/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (new):**

- Scores eval test sets via GitHub Copilot API using the same 8–9 quality dimensions defined in the validate skill.
- No hardcoded model — uses Copilot plan default (works for Individual/Business/Enterprise).
- Scores all cases by default (sample = 100, the eval set cap).
- UTF-8 safe on Windows.

image

**Here's exactly what the script does step by step:**

1. load_topic_context()  → reads modelDescription from every topic .mcs.yml
2. extract_cases()       → reads input + expectedOutput from every eval .mcs.yml
3. build_judge_prompt()  → combines topics + cases + all 8-9 dimension definitions into one big prompt
4. call_judge()          → sends that prompt to Copilot API, gets back structured JSON scores
5. render_report()       → formats the JSON into the readable report

**Quality eval:**

image

**Grouping of topics presented to the user:**

image

### CODE REVIEW

**Comment by amilandi** [comment]
**Re: quality-fix-flow.md — re-validate only edited files + flagged dimensions**

Nit: Coverage and Redundancy are category-level dimensions — they evaluate the spread across the full set, not individual files. With this change to only re-score edited files + flagged dimensions, you might miss cases where a fix introduces new redundancy with an *unedited* file. Consider noting that those two dimensions should still be scored against the full category set even when only specific files were edited.

**Review by amilandi — DISMISSED** [review]

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+            )
+        else:
+            cases_text_parts.append(
+                f"Case {i} ({c['file']}) [single-turn]:\n"
+                f"  input: {c['input']}\n"
+                f"  expectedOutput: {c['expected']}"
+            )
+    cases_text = "\n\n".join(cases_text_parts)
+
+    alignment_schema = '\n    "Topic Alignment": <1-5 or null>,' if include_alignment else ""
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+{{
+  "scores": {{
+    "Validity": <1-5>,
+    "Realism": <1-5>,
+    "Assertion Quality": <1-5>,
+    "Coverage": <1-5>,
+    "Diversity": <1-5>,
+    "Redundancy": <1-5>,
+    "Failure Mode Coverage": <1-5>,
+    "Discriminative Power": <1-5>{alignment_schema}
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py:343]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+                if role and text:
+                    turns.append({"role": role, "text": text})
+            if turns:
+                cases.append({
+                    "file": name,
+                    "kind": "multi-turn",
+                    "turns": turns,
+                    "input": turns[0]["text"] if turns else "",
+                    "expected": "",
+                })
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+    "Coverage": "Cases span a meaningful spread of sub-topics and positive/boundary/negative types",
+    "Diversity": "Inputs use genuinely different vocabulary, structure, and formality levels",
+    "Redundancy": "No two cases test the exact same input and expected behavior",
+    "Failure Mode Coverage": "Negative/edge cases reflect realistic failure modes, not contrived refusals",
+    "Discriminative Power": "Inputs are clearly scoped so they won't accidentally trigger the wrong topic",
+    "Topic Alignment": "Each case matches what its corresponding topic actually does",
+}
+
+# Categories where topic alignment check applies
+TOPIC_ALIGNMENT_CATEGORIES = {"topic-triggering", "integration-data", "multi-turn"}
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+    try:
+        result = subprocess.run(
+            ["gh", "auth", "token"],
+            capture_output=True, text=True, timeout=10
+        )
+        token = result.stdout.strip()
+        if not token:
+            print("ERROR: gh auth token returned empty. Run 'gh auth login' first.", file=sys.stderr)
+            sys.exit(1)
+        return token
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py:201]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md [hunk 6]
```diff
@@ -3,18 +3,47 @@
...
+
+```
+python scripts/evaluate_evals.py --category {category}
+```
+
+Where `{category}` is the subfolder name under `evaluations/` — derive it
+from the file paths you were given. For example, if the files are in
+`evaluations/topic-triggering/`, the category is `topic-triggering`. The
+script calls the GitHub Models API and returns dimension scores and flagged
+cases.
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/quality-fix-flow.md [hunk 7]
```diff
@@ -36,9 +36,11 @@ When the user responds:
...
    > | 2 | `{filename}.mcs.yml` | {dimension} | expectedOutput | `{old output}` | `{new output}` |
 
-4. Re-invoke the validate subagent on all files in the category (not just
-   the edited ones — category-level dimensions need the full set) and display
-   the new quality report before proceeding.
+4. Re-invoke the validate subagent on **only the edited files** and instruct it
+   to re-score only the **dimensions that were flagged** in the original report.
+   Pass both the edited file paths and the list of flagged dimension names so
+   the subagent skips dimensions that weren't flagged and are unaffected by the
+   fixes. Display the updated scores before proceeding.
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/quality-fix-flow.md]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+    try:
+        text = path.read_text(encoding="utf-8")
+    except OSError:
+        return None
+
+    try:
+        import yaml
+        data = yaml.safe_load(text)
+    except ImportError:
+        data = _parse_simple_yaml(text)
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 9]
```diff
@@ -104,7 +132,8 @@ AND contains at least one `.mcs.yml` file. Mark it as **existing** or **missing*
 > | Multi-Turn | {✅ exists (X tests) / ❌ missing} |
 >
 > I can generate the **{M} missing** category/categories ({list missing names}),
-> or regenerate everything from scratch. What would you prefer?
+> generate for a **specific area** (e.g., just Workday topics), or regenerate
+> everything from scratch. What would you prefer?
```
[the remark this answers is not in the document]
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md:136]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+def _parse_simple_yaml(text: str) -> dict:
+    """
+    Minimal YAML parser for the eval file schema.
+    Handles the two known shapes: EvaluationData and MultiTurnEvaluationCase.
+    Falls back to raw text if parsing fails.
+    """
+    try:
+        import yaml
+        return yaml.safe_load(text)
+    except ImportError:
```
Comment (nehaoss) [inline solutions/ess-maker-skills/scripts/evaluate_evals.py]: `_parse_simple_yaml` (lines 97–117) is dead code — it's defined but never called anywhere. `load_eval_file` calls `yaml.safe_load` directly and does `sys.exit(1)` on ImportError, making this fallback unreachable. Either remove it or wire it into `load_eval_file` as the actual fallback path.

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+    if not topics_dir.exists():
+        return context
+
+    for yml_file in sorted(topics_dir.glob("*.mcs.yml")):
+        try:
+            text = yml_file.read_text(encoding="utf-8")
+            try:
+                import yaml
+                data = yaml.safe_load(text)
+            except ImportError:
```
Comment (nehaoss) [inline solutions/ess-maker-skills/scripts/evaluate_evals.py:157]: Inconsistent PyYAML-missing behavior: `load_eval_file` (line 126) does `sys.exit(1)` when yaml is unavailable, but `load_topic_context` here silently swallows the ImportError and sets `data = {}`. This means topic alignment scoring silently degrades to "no topics found" instead of giving the user a clear error. Either both should exit, or both should degrade gracefully with a warning printed to stderr.

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 12]
```diff
@@ -0,0 +1,619 @@
...
+                "\nERROR: GitHub Copilot API returned 401 Unauthorized.\n"
+                "Ensure you have an active GitHub Copilot subscription and run:\n"
+                "  gh auth login\n"
+                f"Details: {err_body}",
+                file=sys.stderr,
+            )
+            sys.exit(1)
+        elif e.code == 429:
+            print("\nERROR: Rate limited by GitHub Models API. Wait a moment and retry.", file=sys.stderr)
+            sys.exit(1)
```
Comment (nehaoss) [inline solutions/ess-maker-skills/scripts/evaluate_evals.py:268]: No retry or backoff on 429 — the script just exits. For a tool that may process multiple categories sequentially (each making an API call), a single transient rate-limit kills the entire run and loses all previously-computed results. Consider adding at least one retry with a short sleep on 429, or saving partial results before exiting.

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md [hunk 13]
```diff
@@ -104,7 +132,8 @@ AND contains at least one `.mcs.yml` file. Mark it as **existing** or **missing*
 > | Multi-Turn | {✅ exists (X tests) / ❌ missing} |
 >
 > I can generate the **{M} missing** category/categories ({list missing names}),
-> or regenerate everything from scratch. What would you prefer?
+> generate for a **specific area** (e.g., just Workday topics), or regenerate
```
Comment (nehaoss) [inline solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: The "SOME categories exist" prompt template hardcodes "just Workday topics" as the example — contradicting the "Do NOT hardcode ServiceNow/Workday/SuccessFactors" instruction two lines above in the "ALL missing" template. Use a placeholder like `{first area label}` here too (same pattern as the ALL-missing block above it) so the agent derives the example from the actual agent's topics.

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 14]
```diff
@@ -0,0 +1,619 @@
...
+                    "You are an expert evaluator of AI agent test sets. "
+                    "Your job is to assess the quality of evaluation test cases "
+                    "for an employee-facing IT support chatbot. "
+                    "Be precise, critical, and actionable. "
+                    "Always respond with valid JSON matching the schema provided."
+                ),
+            },
+            {"role": "user", "content": prompt},
+        ],
+        "max_tokens": 2000,
```
Comment (nehaoss) [inline solutions/ess-maker-skills/scripts/evaluate_evals.py]: `max_tokens: 2000` may be too tight for the JSON response schema when evaluating a large sample (100 cases). The response includes per-dimension scores, up to 5 flagged items with file names and issue descriptions, plus a recommendation. If the model hits the token ceiling, the JSON will be truncated and the `json.loads` call at line 575 will fail with a parse error — silently losing that category's results. Consider bumping to 3000–4000, or at minimum detecting a `finish_reason: "length"` in the response to give a clearer error message than "parse error".

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md [hunk 15]
```diff
@@ -3,18 +3,57 @@
...
+folder. Use the script below to score the files, then add actionable guidance
+for any flagged cases.
 
 ---
 
-## Step 1 — Read the files
+## Step 1 — Run the quality script
+
+Run the following command to score the eval files:
+
```
Comment (nehaoss) [inline solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md:15]: The validate SKILL.md tells the subagent to run `python scripts/evaluate_evals.py --agent {agent-slug} --category {category}` but never specifies a `--topics-dir` argument. When the subagent's CWD is the `solutions/ess-maker-skills/` folder (as the maker kit instructions require), the script's auto-detection looks for `workspace/agents/{slug}/topics/` relative to `scripts/..`, which resolves correctly. However, if invoked from a different CWD (e.g. the repo root), the topics dir won't be found and Topic Alignment silently scores as null. Consider either documenting the required CWD or having the SKILL.md pass `--topics-dir` explicitly.

File: solutions/ess-maker-skills/scripts/evaluate_evals.py (hunk 10 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py (hunk 11 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py:157]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py (hunk 12 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py:268]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md (hunk 13 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/create/SKILL.md]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py (hunk 14 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py]: fixed

File: solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md (hunk 15 above)
Comment (nehaoss) [reply solutions/ess-maker-skills/src/skills/evaluations/validate/SKILL.md:15]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 16]
```diff
@@ -0,0 +1,608 @@
...
+        for f in flagged:
+            flagged_by_dim.setdefault(f["dimension"], []).append(f)
+
+        label = SCORE_LABELS.get(overall, "")
+        print(f"  Overall quality: {overall}/5  {label}")
+        print()
+
+        for dim, score in scores.items():
+            if score is None:
+                continue
```
Comment (GrahamMcMynn) [inline solutions/ess-maker-skills/scripts/evaluate_evals.py:415]: `render_report` silently drops flagged cases whose `dimension` either (a) was scored `null` (the `if score is None: continue` above skips the dimension entirely, including any flagged cases attached to it) or (b) doesn't appear in `scores` at all (e.g. a dimension name the judge mis-spelled or invented). The flag gets into `flagged_by_dim` but is never printed.

Since `flagged` is the only surface for case-level issues in the report, those cases vanish without warning. Consider adding a tail loop after the per-dimension rendering that prints any `flagged_by_dim` entries not yet shown (e.g. under an "Other flagged cases" subhead), so flagged items under null-scored or unmatched dimensions still surface.

Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py:415]: fixed

File: solutions/ess-maker-skills/scripts/evaluate_evals.py [hunk 17]
```diff
@@ -0,0 +1,618 @@
...
+
+            model_desc = data.get("modelDescription", "") or ""
+            model_desc = model_desc.strip()
+            if not model_desc:
+                continue
+
+            context[yml_file.stem] = model_desc
+        except Exception:
+            continue
```
Comment (GrahamMcMynn) [inline solutions/ess-maker-skills/scripts/evaluate_evals.py:176]: `load_topic_context` wraps the per-file block in `except Exception: continue` (line 172–173), so a topic YAML with a parse error, encoding issue, or unexpected shape is silently skipped — the user only sees a lower-than-expected `Loaded N topic definitions` count (or no message at all if `N == 0`).

`load_eval_file` above doesn't catch `yaml.YAMLError` at all, so a malformed eval file crashes the script with a traceback. After the earlier fix made the PyYAML-missing path consistent (`sys.exit(1)` in both), this is the remaining asymmetry.

Suggestion: in `load_topic_context`, narrow the `except` to `(yaml.YAMLError, OSError)` and print a one-line warning to stderr (e.g. `WARNING: could not parse {yml_file.name} for topic alignment: {e}`) before `continue`. Users get a diagnostic when alignment context silently degrades, and unexpected exceptions still propagate.

Comment (nehaoss) [reply solutions/ess-maker-skills/scripts/evaluate_evals.py:176]: fixed

**Review by GrahamMcMynn — APPROVED** [review]
